browser/ui/views/toolbar/ai_chat_button.cc

```diff
@@ -0,0 +1,83 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/toolbar/ai_chat_button.h"
+
+#include <memory>
+
+#include "brave/app/brave_command_ids.h"
+#include "brave/components/ai_chat/core/common/pref_names.h"
+#include "brave/components/vector_icons/vector_icons.h"
+#include "brave/grit/brave_generated_resources.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_commands.h"
+#include "components/prefs/pref_service.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/base/metadata/metadata_impl_macros.h"
+#include "ui/base/models/simple_menu_model.h"
+#include "ui/views/accessibility/view_accessibility.h"
+#include "ui/views/controls/button/button_controller.h"
+
+namespace {
+
+class AIChatButtonMenuModel : public ui::SimpleMenuModel,
+                              public ui::SimpleMenuModel::Delegate {
+ public:
+  explicit AIChatButtonMenuModel(PrefService* prefs)
+      : ui::SimpleMenuModel(this), prefs_(*prefs) {
+    Build();
+  }
+
+  AIChatButtonMenuModel(const AIChatButtonMenuModel&) = delete;
+  AIChatButtonMenuModel& operator=(const AIChatButtonMenuModel&) = delete;
+  ~AIChatButtonMenuModel() override = default;
+
+ private:
+  enum ContextMenuCommand { kHideAIChatButton };
+
+  void Build() {
+    AddItemWithStringId(ContextMenuCommand::kHideAIChatButton,
+                        IDS_HIDE_SIDE_PANEL_TOOLBAR_BUTTON);
+  }
+
+  // ui::SimpleMenuModel::Delegate:
+  void ExecuteCommand(int command_id, int event_flags) override {
+    if (command_id == ContextMenuCommand::kHideAIChatButton) {
+      prefs_->SetBoolean(ai_chat::prefs::kBraveAIChatShowToolbarButton, false);
+    }
+  }
+
+  raw_ref<PrefService> prefs_;
+};
+
+}  // namespace
+
+AIChatButton::AIChatButton(Browser* browser)
+    : ToolbarButton(base::BindRepeating(&AIChatButton::ButtonPressed,
+                                        base::Unretained(this))),
+      browser_(*browser) {
+  auto* prefs = browser->profile()->GetOriginalProfile()->GetPrefs();
+  SetMenuModel(std::make_unique<AIChatButtonMenuModel>(prefs));
+  SetVectorIcon(kLeoProductBraveLeoIcon);
+
+  // Visibility is managed by |ToolbarView|.
+  SetVisible(false);
+
+  SetTooltipText(l10n_util::GetStringUTF16(IDS_TOOLTIP_SIDEBAR_SHOW));
+  set_context_menu_controller(this);
+  button_controller()->set_notify_action(
+      views::ButtonController::NotifyAction::kOnPress);
```

Comment:
Q. I'm curious why we need this.

Reply:
It's a good question because I don't think there's a document or comment explaining when in Chromium to use either `kOnPress` or `kOnRelease`. Chromium seems to use both unpredictably. The sidebar button uses `kOnPress` and since this also toggles the sidebar, I copied that. However, the sidebar button "looks" more like a toggle since it highlights too. I suppose I'll remove it for now, like the other toolbar buttons.